addons/sale_stock/models/sale_order.py

```diff
@@ -64,11 +58,32 @@ def _compute_expected_date(self):
                 expected_date = min(dates_list) if order.picking_policy == 'direct' else max(dates_list)
                 order.expected_date = fields.Datetime.to_string(expected_date)
 
-    @api.model
-    def create(self, vals):
-        if 'warehouse_id' not in vals and 'company_id' in vals and vals.get('company_id') != self.env.company.id:
-            vals['warehouse_id'] = self.env['stock.warehouse'].search([('company_id', '=', vals.get('company_id'))], limit=1).id
-        return super(SaleOrder, self).create(vals)
+    @api.depends('user_id','company_id')
+    def _compute_warehouse_id(self):
+        for order in self:
+            company = order.company_id.id
+            if order.warehouse_id:
+                if order.warehouse_id.company_id.id != company:
+                    order.warehouse_id = False
+            if not order.warehouse_id:
+                warehouse_id = order.env['stock.warehouse'].search([('company_id', '=', company)], limit=1)
+                order.warehouse_id = warehouse_id.id
+                if order.user_id:
+                    if order.user_id.with_company(company).property_warehouse_id:
+                        order.warehouse_id = order.user_id.with_company(company).property_warehouse_id.id
+
+    @api.onchange('user_id')
+    def onchange_user_id(self):
```

Comment:
This is conflicting with the compute method (and should disappear as we don't want to introduce business code into onchange methods).

Reply:
Note: Know that order._origin returns the real record representation.